### PR TITLE
proof of concept for wire-backend

### PR DESCRIPTION
**Pull request by fisx**
TODO:
- [x] create some sample entries in ldap
- [x] pull them
- [x] convert them into scim
- [x] post everything to a scim peer

update and deletion will happen in a different PR: #2

### CODE REVIEW

**Comment by fisx** [comment]
```

*LdapScimBridge> :main "sample-conf.yaml"
*** Exception: FailureResponse (Request {requestPath = (BaseUrl {baseUrlScheme = Http, baseUrlHost = "localhost", baseUrlPort = 8088, baseUrlPath = "/scim/v2"},"/Users"), requestQueryString = fromList [], requestBody = Just ((),application/scim+json;charset=utf-8), requestAccept = fromList [application/scim+json;charset=utf-8,application/scim+json,application/json;charset=utf-8,application/json], requestHeaders = fromList [("Authorization","<REDACTED>")]), requestHttpVersion = HTTP/1.1, requestMethod = "POST"} (Response {responseStatusCode = Status {statusCode = 400, statusMessage = "Bad Request"}, responseHeaders = fromList [("Transfer-Encoding","chunked"),("Date","Wed, 11 Aug 2021 11:02:28 GMT"),("Server","Warp/3.3.13"),("Cache-Control","no-cache, no-store"),("Pragma","no-cache"),("Content-Type","application/scim+json;charset=utf-8")], responseHttpVersion = HTTP/1.1, responseBody = "{\"scimType\":\"invalidValue\",\"status\":\"400\",\"schemas\":[\"urn:ietf:params:scim:api:messages:2.0:Error\"],\"detail\":\"externalId must be a valid email address or (if there is a SAML IdP) a valid SAML NameID\"}"})
```

so i can pull ldap data, convert it to scim, post it, and get an error that indicates that the post has been parsed correctly, and i just need to work harder on my mapping.  success!  :)

File: src/LdapScimBridge.hs [hunk 1]
```diff
@@ -1,7 +1,227 @@
...
+scimUpdate = ScimSite.usPutUser scimRoutes
+
+scimDelete :: ScimServer.Id -> IO NoContent
+scimDelete = ScimSite.usDeleteUser scimRoutes
+
+----------------------------------------------------------------------
+
+main :: IO ()
+main = do
+  myconf :: SearchConf <- ByteString.readFile "./sample-conf.yaml" >>= either (error . show) pure . Yaml.decodeEither'
```
[the remark this answers is not in the document]
Comment (fisx) [reply src/LdapScimBridge.hs]: done.

File: README.md [hunk 2]
```diff
@@ -6,4 +6,67 @@
...
+this code base, but in the same repo under `/examples/wire.com/`, and
+add a field to yaml that points to the downloaded csv file and the
+column with the ID for deletion information for all scim peers that do
+not implement "get all users" requests.
+
+BETTER IDEA (thanks julia): Ad uses a deleted objects folder, so this
+makes sense to use that.
+https://www.lepide.com/how-to/restore-deleted-objects-in-active-directory.html.
+
+do we need to change the `Mock` tag for `Scim.User.User` to something
```
[the remark this answers is not in the document]
Comment (fisx) [reply README.md:25]: i think `Mock` is enough for our purposes.  when i clean up the readme (before i declare this to be ready for shipping), i will remove this and add a github issue for future work.

File: src/LdapScimBridge.hs [hunk 3]
```diff
@@ -1,7 +1,227 @@
...
+  ScimSchema.empty schemas (error "undefined") ScimSchema.NoUserExtra
+  where
+    schemas = [Scim.User20]
+
+ldapToScim ::
+  forall scim.
+  scim ~ Either [MappingError] (ScimSchema.User ScimServer.Mock) =>
+  SearchConf ->
+  SearchEntry ->
+  scim
```
[the remark this answers is not in the document]
Comment (fisx) [reply src/LdapScimBridge.hs]: i should just give up on this pattern, people keep tripping over it.  :)

no reason, except i think this is easier to read.  should i change it?

File: sample-conf.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  password: "geheim hoch drei"
+  base: "ou=People,dc=nodomain"
+  objectClass: "account"
+  codec: "utf8"
+scimTarget:
+  tls: false
+  host: "localhost"
+  port: 8088
+  path: "/scim/v2"
+  token: "Bearer scim-test-token/team=471a5f6e-86fa-4850-b6e0-9758259afcdf/code=b873fbf9-e3b5-487b-8098-ab1f47fed7b0"
```
Comment (fisx) [inline sample-conf.yaml:15]: I'm torn between moving the `"Bearer "` part to the code and leaving it here.  @jmatsushita any opinion?

File: src/LdapScimBridge.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmatsushita) [reply src/LdapScimBridge.hs]: I mean it is more general, but if you don't use the generality, less readable. No need to change it.

**Pull request merged**